pennylane/measurements/shots.py

```diff
@@ -0,0 +1,134 @@
+# Copyright 2018-2023 Xanadu Quantum Technologies Inc.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""This module contains the Shots class to hold shot-related information."""
+
+from collections import namedtuple
+from collections.abc import Sequence as SequenceType
+from functools import singledispatchmethod
+from typing import Sequence, Tuple, Union
+import numpy as np
+
+ShotCopies = namedtuple("ShotCopies", ["shots", "copies"])
+"""A namedtuple that represents a shot quantity being repeated some number of times."""
+
+
+def valid_int(s):
+    """Returns True if s is an non-negative integer."""
+    return isinstance(s, int) and s > 0
+
+
+def valid_tuple(s):
+    """Returns True if s is a tuple of the form (shots, copies)."""
+    return isinstance(s, tuple) and len(s) == 2 and valid_int(s[0]) and valid_int(s[1])
+
+
+class Shots:
+    """A data class that stores shot information."""
+
+    total_shots: int = None
+    """The total number of shots to be executed."""
+
+    shot_vector: Tuple[ShotCopies] = None
+    """The list of ShotCopies to be executed. Each element is of the form (shots, copies)."""
+
+    _SHOT_ERROR = ValueError(
+        "Shots must be a single non-negative integer, a tuple pair of the form (shots, counts), or a sequence of these types."
+    )
+
+    _frozen = False
+
+    def __new__(cls, shots):
+        return shots if isinstance(shots, cls) else object.__new__(cls)
+
+    def __setattr__(self, name, value):
+        if self._frozen:
+            raise AttributeError(
+                "Shots is an immutable class. Consider creating a new instance if you need different shot values."
+            )
+        return super().__setattr__(name, value)
+
+    @singledispatchmethod
+    def __init__(self, shots):
+        if isinstance(shots, self.__class__):
+            return  # self already _is_ shots as defined by __new__
+        raise self._SHOT_ERROR
+
+    @__init__.register
+    def __int_init__(self, shots: int):
+        if shots < 1:
+            raise self._SHOT_ERROR
+        self.total_shots = shots
+        self.shot_vector = (ShotCopies(shots, 1),)
+        self._frozen = True
+
+    @__init__.register
+    def __None_init__(self, shots: type(None)):  # pylint:disable=unused-argument
+        self.total_shots = None
+        self.shot_vector = ()
+        self._frozen = True
+
+    @__init__.register
+    def __tuple_init__(self, shots: tuple):
```

Comment:
What about `shots=((10,2), (20,2))`, a tuple of shot copies?

Reply:
imo i'd rather reject it to keep the dispatching simpler. you can just give `[(10, 2), (20, 2)]`. I can update the `_SHOT_ERROR` to say "a list of these types" instead of a sequence. I can also add support for a tuple of possible values, it just kinda takes away from the added value of single-dispatching

Reply:
`Shots((1,2))` would be first one shot, then two shots.  If someone wanted "one shot twice", they would do `Shots( ((1,2), ) )`. An iterable with one shot tuple.


Compare to current behaviour.
```
>>> dev = qml.device('default.qubit', wires=2, shots=(1,2))
>>> dev.shots
3
>>> dev.shot_vector
(3, [ShotTuple(shots=1, copies=1), ShotTuple(shots=2, copies=1)])
>>> dev2 = qml.device('default.qubit', wires=2, shots=((1,2,), ))
>>> dev2.shots
2
>>> dev2.shot_vector
[ShotTuple(shots=1, copies=2)]
```

Reply:
Also, I think it's important that we don't distinguish between different types of iterable (list, tuple) for what we accept.
Permissive on user-inputs, restrictive on internals.

Reply:
updated how we handle various iterables and added examples to docstring